### PR TITLE
Redux Part One

### PR DESCRIPTION
**Pull request by NicholasTD07**
## Why did you write this blog?

When I watched the 2 minutes short videos, [Getting Started with Redux](https://egghead.io/series/getting-started-with-redux) by [Dan Abramov (on GitHub)](https://github.com/gaearon),

I was amazed by how simple Redux, as a concept and also how simple it can be implemented in JavaScript.

I knew I would write Redux in Swift one day and maybe share about the experience I had with building Redux. **This is it.**
## You know there is [ReSwift](https://github.com/ReSwift/ReSwift) Right?

For people who don't know ReSwift, it is a Redux like framework written in Swift.

To the Q: YES. I also know [Karl Bowden (on GitHub)](https://github.com/agentk) who's one of the main contributors behind ReSwift.
He gave me the inspiration for writing my own version of ReSwift because it's very easy to roll your own version, 50 lines of code. BANG!

### CODE REVIEW

File: drafts/swift/redux-in-swift-pt-1.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+> Redux architecture revolves around a **strict unidirectional data flow**.
+
+![](https://www.dropbox.com/s/51evjhxw35t59x2/Redux.png?raw=1)
+
+[Redux Doc: Data Flow](http://redux.js.org/docs/basics/DataFlow.html)
+
+### WAT?
+
+In essense, it's an idea which trys to
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *essence, *tries

File: drafts/swift/redux-in-swift-pt-1.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+- seperate/isolate where changes can happen to the "State" of an application
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: I understand what you mean by this line, but I wonder if it could be made clearer: would this communicate the same thing?

> seperate/isolate the 'state' of an application to a single location

File: drafts/swift/redux-in-swift-pt-1.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+- limits the flow of the data in an application
+
+### What Composes Redux?
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: > What is Redux composed of?

or maybe just

> What is Redux?

File: drafts/swift/redux-in-swift-pt-1.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+Five types (of things) composes Redux:
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: This is an artefact of the English language: Redux is composed _of_ some things, rather than those things _composing_ Redux.*

_*Actually, both are correct, but the way you've written it above is a bit more academic and less easy to parse for an average reader, I'd say._

File: drafts/swift/redux-in-swift-pt-1.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+- *State*, which could be anything
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: Might be worth making clearer what this state actually is, i.e the (entire) internal state of your application.

File: drafts/swift/redux-in-swift-pt-1.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+The `state` is marked as `private(set)` because:
+
+- the *State* in a *Store* should be read-only and
+- "The only way to change the state is to emit an action, an object describing what happened."
+
+[Redux Doc: Three Principles](http://redux.js.org/docs/introduction/ThreePrinciples.html)
+
+### Reducer and Action
+
+A *Store* need a *Reducer* to generate new *State*s from the current *State* and an Action, which means:
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *needs

File: drafts/swift/redux-in-swift-pt-1.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+- A *Store* need to hold onto a *Reducer*
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *needs

File: drafts/swift/redux-in-swift-pt-1.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+- A *Reducer*
+    - takes a *State* and a *Action* and
+    - returns a new *State*
+- An *Action* describes what happen**ed** which needs to differentiate from each other while it may carry some data
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: I'm not sure I understand this sentence - can you rephrase?

File: drafts/swift/redux-in-swift-pt-1.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    init(with reducer: Reducer) {
+        self.reducer = reducer
+    }
+}
+```
+
+#### Things to Note about Reducers
+
+- They should be pure: No side-effects. No API calls. etc.
+- They take `nil` as `State` which implies the default initial state is defined in a *Reducer* rather than a *Store*.
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *They _can_ take

or

*They take an Optional `State`, which means `State` can be `nil`, implying...

File: drafts/swift/redux-in-swift-pt-1.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+    - This is partially why we need that `!` when defining the `state` in the `Store`.
+- They should return the current *State* if they are given an *Action* they cannot handle.
+
+### Dispatching Actions and Notifying Subscribers
+
+An *Action* can be dispatched to a *Store* to change its *State*. After a *Store*'s *State* is changed, the *Store* should notify all of its *Subscribers*. Thus,
+
+- A *Subscriber* is a function takes a *Store*
+- *Subscribers* can subscribe to a *Store* (will be explained in the next section)
+- A *Store* need to hold onto an array of *Subscribers*
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *needs

File: drafts/swift/redux-in-swift-pt-1.md [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+- After a *Store*'s *State* changes, the *Store* need to notify the change to all of its *Subscribers*
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *needs

File: drafts/swift/redux-in-swift-pt-1.md [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+    final func dispatch(action: ActionType) {
+        self.state = reducer(state: state, action: action)
+        subscribers.forEach {
+            $0(store: self)
+        }
+    }
+}
+```
+
+Now the generic class `Store` is almost usable except a few pieces missing. Example usage:
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *except for a few missing pieces

File: drafts/swift/redux-in-swift-pt-1.md [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+
+```swift
+struct Increase: ActionType { }
+
+let counterStore = Store<Int>.init { (state: Int?, action: ActionType) -> Int in /* ... */ }
+
+counterStore.dispatch(Increase())
+```
+
+If you want to run above snippet with minimum change, replace the `/* ... */` with `return 0`.
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *If you want to run the snippet above at this point, you'll need to replace the...

File: drafts/swift/redux-in-swift-pt-1.md [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+
+#### How does a Store Get its Initial State?
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: inconsistent capitalisation: maybe go with

> How does a Store get its initial State

to communicate the specialised meaning of Store and State?

File: drafts/swift/redux-in-swift-pt-1.md [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+So when a *Store* initializes, the following would happen regarding its `state`:
+
+- the `Store` will call its `dispatch` with the `InitialAction`
+- `dispatch` would update the `Store`'s `state` by calling its `reducer` with `nil` as the `state` param and `InitialAction` 
+- the `state` in the `Store` would be set to whatever the `Reducer` returns as the initial state
+    - because a `Reducer` returns the actual initial state when it receives a `nil` as its `state` param
+
+#### The BANG`!`
+
+As you can see from the above and actually running the snippet, the forced unwrapped `state` in the `Store` will always be set to non nil values before it's used as a non nil value. It is not only safe but also convenient to use (no need for `guard let`s everywhere).
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *force-unwrapped

Comment (sam-w) [reply drafts/swift/redux-in-swift-pt-1.md]: *be set to a non-nil value

File: drafts/swift/redux-in-swift-pt-1.md [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+
+### Subscribing to a Store
+
+*Subscriber*s are functions which takes a *Store* as its only param and they can subscribe to *Store*s by calling the following method. When it's subscribed to a *Store*, it should get the *Store*'s current *State*.
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *...as its only param. They can...

Comment (sam-w) [reply drafts/swift/redux-in-swift-pt-1.md]: _...by calling the `subscribe` method as illustrated below. When a *Subscriber_ is subscribed to a _Store_, it will receive the _Store_'s current _State_.

File: drafts/swift/redux-in-swift-pt-1.md [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+```swift
+var counter: Int = -1000
+counterStore.subscribe { (store: Store<Int>) in
+    counter = store.state
+}
+```
+
+#### TESTS!
+
+If you remove all the example snippet and add the following snippet and run the file either
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *example snippets, add the following and run in either:...

File: drafts/swift/redux-in-swift-pt-1.md [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+
+- in Xcode Playground or,
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: - no 'in'

File: drafts/swift/redux-in-swift-pt-1.md [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+- in commond line by running `swift path/to/your-file` or,
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *command

File: drafts/swift/redux-in-swift-pt-1.md [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+- in Vim by doing `:w | !swift %` (personal perference)
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: *preference

File: drafts/swift/redux-in-swift-pt-1.md [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+
+the file should run without any problem which indicates all the assert (tests) passes.
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: 'File' here isn't the most descriptive term - maybe 'test suite' instead?

File: drafts/swift/redux-in-swift-pt-1.md [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+// when state is updated, subscirbers will get notified of the new state
+counterStore.dispatch(CounterActions.Increase)
+assert(counter == counterStore.state)
+```
+
+## What's Next?
+
+In the next blog, I will
+
+- Build a Todo with Swift and the micro Redux framework we built in this blog post.
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: What does 'Build a todo' mean?

File: drafts/swift/redux-in-swift-pt-1.md [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+- explain how to split large *Reducer*s into smaller ones and combine them back into one using something new called *`combineReducers`*.
+    - (TBH: I have no idea what it would be right now. I guess it would be a function like the *Reducers* in this blog)
```
Comment (sam-w) [inline drafts/swift/redux-in-swift-pt-1.md]: Is this line directed at us, or at the reader?

**Review by sam-w — COMMENTED** [review]
Hey, nice work! I've done a quick review of the writing style of the article - I haven't attempted to understand any of the code yet though :)

File: drafts/swift/redux-in-swift-pt-1.md (hunk 11 above)
Comment (NicholasTD07) [reply drafts/swift/redux-in-swift-pt-1.md]: Sam has so many `needs`. 😆 Just kidding.

Comment (NicholasTD07) [reply drafts/swift/redux-in-swift-pt-1.md]: Serious question, when would a native speaker use `need` (like `can`) in third person singular form _as a verb_? (EDITED)

File: drafts/swift/redux-in-swift-pt-1.md (hunk 23 above)
Comment (NicholasTD07) [reply drafts/swift/redux-in-swift-pt-1.md]: @sam-w At the reader. Let me know what you think :)

File: drafts/swift/redux-in-swift-pt-1.md (hunk 11 above)
Comment (agentk) [reply drafts/swift/redux-in-swift-pt-1.md]: Needs vs need. The plural version (needs) is used if anything can happen more than once, or referring to multiple 'need's. Ie, Food is a basic need. A human has basic needs. Food is one of the basic needs.

Most of the uses referring to the Store and Actions, will be in the plural form (needs) as the the Store has State**s**, and Action**s** etc.

File: drafts/swift/redux-in-swift-pt-1.md (hunk 8 above)
Comment (agentk) [reply drafts/swift/redux-in-swift-pt-1.md]: - An _Action_ describes an event that has happen**ed**. Because of this, Action's are best written in the past tense. It's simple to get started if you use the words Did or Was in the Action name. Ie, UserDidPressEject or CompactDiscWasEjected.
  - Action's may also be composed of data about the Event.
  - In that was they can also be thought of like an accounting ledger. You are making a record of an event, that changes the account balances and cannot be changed afterwards. To 'undo' an Action, another action must be submitted that has the equal and opposite effect.
  - Ie, ChequeWasDeposited(value: 100.00), ChequeDidNotClear(value: -100.00)

**Comment by NicholasTD07** [comment]
Done

> One more thing: update the diagram to use a different color for Subscriber.

Done 

```
Some TODOs for myself as a mental note:

- Mention the fact that I am writing Swift 2.2 rather than trendy Swift 3
- Short description of what will be walked through in the blog (Double check whether I have done that)
- How to use/play with the code snippet in the blog
- Update What's Next section with improvements
  - Try to remove the type cast in the reducer function for counterStore
    - Possibly with `combineReducers` and a `defaultReducer` which does all the type casting?
  - from @agentk's suggestions
    - Try to remove the forced-unwrap
    - Upgrade it to Swift 3
    - handle possible strong reference cycle between `Store` and `Subscriber` because of `dispatch`
```
